### DISCUSSION TITLE
Import users without losing ids

### DISCUSSION THREAD

**Discussion by olivierboudet** [keycloak/keycloak#8624] [Ideas]
Hello,
I have a use case where I need to import users without losing their current ids in a keycloak instance where the realm already exists (because the configuration is provided by terraform).

I tested partialImport, it generates new ids.
I tested full import, it does not import users if realm is not created during import (using IGNORE_EXISTING does not work either because the realm is not created).

Ideas : 
- would it be possible to remove the restriction to not import users if the realm already exists ?
- perhaps we can add a new parameter migration `keycloak.migration.forceImportUsers` when we want to import users using an existing realm ?
- would it be possible to add a parameter to partialImport to explicitly ask to not generate new ids ?

**Comment by magoogli** [comment]
I second this - in my use case I am migrating from an existing identity provider to Keycloak, and it is very important for me that each migrated user retains the same id. At the moment I generate a json file of users from my existing identity provider in the format keycloak needs, and then have to do a 'FULL' keycloak import which I find pretty cumbersome. It would be great if the Partial import simply used the Ids for the users which I supply in the file.

In my case simply removing this line in [UsersPartialImport.java](https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/services/src/main/java/org/keycloak/partialimport/UsersPartialImport.java#L113) would solve my trouble, but I don't have enough experience with the project to know what the knock on effects would be. Can anyone provide some background on why that line is there in the first place?

Here is a very similar discussion I found on the [keycloak group](https://keycloak.discourse.group/t/user-uuid-changes-on-import/2192)

**Comment by thebestgin** [comment]
It is not understandable at all why a new UserId is generated during an import. 
In my opinion this is a bug.

Every system that refers to the UserId is broken.

In the end we are talking about uuids. By the way, these are unique.

**Comment by zingmane** [comment]
> It is not understandable at all why a new UserId is generated during an import.
> In my opinion this is a bug.

I agree, for me too.
Every time a import groups or realm roles I loose the relationships between groups and users respectively groups and realms though the UUID is provided in the JSON.

Wouldn't be an upsert way much better? If UUIDs are provided and in keycloak exist they should be updated, else they should be created. 
For creation/insert, it would be fine if the uuids were not used, since there is no way to know from the outside if the IDs exist.

**Comment by croissong** [comment]
FYI, the following works & keeps the original user ids:
1. create realm export
```sh
kc.sh export --dir /tmp/realm-export/ --realm <realm> --users same_file
```
2. use partialImport on the users json
```sh
kcadm.sh create partialImport --realm <realm> -s ifResourceExists=SKIP -o -f <realm>-users-0.json
```
(or use `ifResourceExists=FAIL` or `ifResourceExists=OVERWRITE`)

**Reply by ManiBharathi-R** [reply]
@croissong , can you please mention which Keycloak version you tried the above command with?

**Reply by croissong** [reply]
@ManiBharathi-R Iirc i ran the export command in v15 and the import in v25

**Reply by clementgineste** [reply]
> FYI, the following works & keeps the original user ids:
> 
> 1. create realm export
> 
> ```shell
> kc.sh export --dir /tmp/realm-export/ --realm <realm> --users same_file
> ```
> 
> 2. use partialImport on the users json
> 
> ```shell
> kcadm.sh create partialImport --realm <realm> -s ifResourceExists=SKIP -o -f <realm>-users-0.json
> ```
> 
> (or use `ifResourceExists=FAIL` or `ifResourceExists=OVERWRITE`)

Not working for me here 

KC 21.0.2

```
root@keycloak:/opt/backups/temp# /opt/keycloak/bin/kcadm.sh create partialImport --realm oclock -s ifResourceExists=SKIP -o -f oclock-users-xxx.json 
{
  "overwritten" : 0,
  "added" : 1,
  "skipped" : 0,
  "results" : [ {
    "action" : "ADDED",
    "resourceType" : "USER",
    "resourceName" : "arthur-xxx@gmail.com",
    "id" : "9613d654-9a29-43a2-babb-36c08830ee2e"
  } ]
}
```

But that's not the ID in export json :(

edit: `kc.sh import -dir ..` worked and old ID came back !
